### PR TITLE
#20311: Fix association race condition and add reproducer

### PR DESCRIPTION
**Pull request by drewhk**

### CODE REVIEW

File: akka-remote/src/main/scala/akka/remote/artery/ArterySubsystem.scala [hunk 1]
```diff
@@ -67,30 +67,27 @@ private[remote] class ArterySubsystem(_system: ExtendedActorSystem, _provider: R
...
-      }
+      else associate(remoteAddress)
 
     association.send(message, senderOption, recipient)
   }
 
-  private def getAssociation(remoteAddress: Address): Association = {
+  private def associate(remoteAddress: Address): Association = {
     val current = associations.get(remoteAddress)
     if (current ne null) current
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/ArterySubsystem.scala]: we must also, perhaps later, think about how to handle races when removing associations

Comment (drewhk) [inline akka-remote/src/main/scala/akka/remote/artery/ArterySubsystem.scala]: yes, some garbage collection will be needed.

File: akka-remote/src/main/scala/akka/remote/artery/ArterySubsystem.scala [hunk 2]
```diff
@@ -67,30 +67,27 @@ private[remote] class ArterySubsystem(_system: ExtendedActorSystem, _provider: R
...
     else {
-      val newAssociation = new Association(materializer, remoteAddress, transport)
-      val currentAssociation = associations.putIfAbsent(remoteAddress, newAssociation)
-      if (currentAssociation eq null) {
-        newAssociation
-      } else currentAssociation
+      associations.computeIfAbsent(remoteAddress, new java.util.function.Function[Address, Association] {
```
Comment (viktorklang) [inline akka-remote/src/main/scala/akka/remote/artery/ArterySubsystem.scala]: «The default implementation may retry these steps when multiple threads attempt updates including potentially calling the mapping function multiple times.»

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentMap.html#computeIfAbsent-K-java.util.function.Function-

Comment (drewhk) [inline akka-remote/src/main/scala/akka/remote/artery/ArterySubsystem.scala]: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent-K-java.util.function.Function-

> The entire method invocation is performed atomically, so the function is applied at most once per key. Some attempted update operations on this map by other threads may be blocked while computation is in progress, so the computation should be short and simple, and must not attempt to update any other mappings of this map.

Comment (viktorklang) [inline akka-remote/src/main/scala/akka/remote/artery/ArterySubsystem.scala]: Doesn't `newAssociation.associate() // This is a bit costly for this blocking method :(` violate `short and simple`?

Comment (drewhk) [inline akka-remote/src/main/scala/akka/remote/artery/ArterySubsystem.scala]: Yes it does, hence the comment. This is a skeleton/scaffolding that we currently work with, this part is a simple hack that just happens to work for now.

Comment (viktorklang) [inline akka-remote/src/main/scala/akka/remote/artery/ArterySubsystem.scala]: Ah, so it should probably have a FIXME attached

Comment (drewhk) [inline akka-remote/src/main/scala/akka/remote/artery/ArterySubsystem.scala]: Well, currently the whole code is a big FIXME on this branch :) This branch is so far for experimentation and seeing how things fit or not fit together.

Comment (viktorklang) [inline akka-remote/src/main/scala/akka/remote/artery/ArterySubsystem.scala]: Agreed. In my experience, if one does not clearly demarcate one's FIXMEs and do not scrap all the code at some point prior to merging, they tend to leak into the codebase—so faster and cheaper to annotate FIXMEs early and often :)

Comment (drewhk) [inline akka-remote/src/main/scala/akka/remote/artery/ArterySubsystem.scala]: It does not hurt for sure to add one more here. I will do merge anyway though and fix it in the next round as PR validation is rather costly :(

Comment (drewhk) [inline akka-remote/src/main/scala/akka/remote/artery/ArterySubsystem.scala]: Well, I know how to fix this so maybe I should just do it.

Comment (viktorklang) [inline akka-remote/src/main/scala/akka/remote/artery/ArterySubsystem.scala]: Cool, thanks!

**Comment by patriknw** [comment]
LGTM, please merge if ready

**Comment by drewhk** [comment]
I merge this and fix the remaining issues later as we need this fix

**Pull request merged**